c2rust-analyze/src/main.rs

```diff
@@ -456,6 +458,65 @@ fn label_rvalue_tys<'tcx>(acx: &mut AnalysisCtxt<'_, 'tcx>, mir: &Body<'tcx>) {
     }
 }
 
+/// Set flags in `acx.ptr_info` based on analysis of the `mir`.  This is used for `PointerInfo`
+/// flags that represent non-local properties or other properties that can't be set easily when the
+/// `PointerId` is first allocated.
+fn update_pointer_info<'tcx>(acx: &mut AnalysisCtxt<'_, 'tcx>, mir: &Body<'tcx>) {
+    let mut write_count = HashMap::with_capacity(mir.local_decls.len());
+    let mut rhs_is_ref = HashSet::new();
+
+    for (bb, bb_data) in mir.basic_blocks().iter_enumerated() {
+        for (i, stmt) in bb_data.statements.iter().enumerate() {
+            let (pl, rv) = match &stmt.kind {
+                StatementKind::Assign(x) => &**x,
+                _ => continue,
+            };
+
+            eprintln!(
+                "update_pointer_info: visit assignment: {:?}[{}]: {:?}",
+                bb, i, stmt
+            );
+
+            // Note we ignore `c_void_casts` here.  It shouldn't affect any of the patterns we're
+            // looking for.
+
+            if !pl.is_indirect() {
+                // This is a write directly to `pl.local`.
+                *write_count.entry(pl.local).or_insert(0) += 1;
+                eprintln!("  record write to LHS {:?}", pl);
+            }
+
+            let ref_pl = match *rv {
+                Rvalue::Ref(_rg, _kind, pl) => Some(pl),
+                Rvalue::AddressOf(_mutbl, pl) => Some(pl),
+                _ => None,
+            };
+            if let Some(ref_pl) = ref_pl {
+                // For simplicity, we consider taking the address of a local to be a write.  We
+                // expect this not to happen for the sorts of temporary refs we're looking for.
+                if !ref_pl.is_indirect() {
+                    eprintln!("  record write to ref target {:?}", ref_pl);
+                    *write_count.entry(ref_pl.local).or_insert(0) += 1;
+                }
+
+                rhs_is_ref.insert(pl.local);
+            }
+        }
+    }
+
+    for local in mir.local_decls.indices() {
+        let is_temp_ref = mir.local_kind(local) == LocalKind::Temp
+            && write_count.get(&local).copied().unwrap_or(0) == 1
```

Comment:
am i reading this right that you're only looking for cases where an address of a local is taken but not a reborrow?

Reply:
No, we don't care what `Place` the temporary reference points to.  `_1 = &_2`, `_1 = &_2.0`, and `_1 = &*_2` are all treated the same for our purposes.  We only care whether the LHS variable looks like it's a temporary storing the result of a source-level `&...` expression.